### ISSUE TITLE
words: Unbound variable nickname in IRCUser for invalid user name

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5780]
|<img alt="j2a's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| j2a reported|
|-|-|
|Trac ID|trac#5780|
|Type|defect|
|Created|2012-07-14 23:06:23Z|
|Branch|https://github.com/twisted/twisted/tree/unbound-nick-5780|

If user send NICK command with non-utf8 non-ascii symbols, IRCUser raises UnboundLocalError:

```exceptions.UnboundLocalError: local variable 'nickname' referenced before assignment```

patch and test is attached.

Attachments:

* [unbound_local_error.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7a4/7a422879511b7e2652ca8839f9042a62a431e596/a1e5cde2c02862886e07b394fa00129d8b8fbfc6.patch) (1776 bytes) - added by j2a on 2012-07-14 23:07:13Z - Fix for UnboundLocalError and test for it
* [5780v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7a4/7a422879511b7e2652ca8839f9042a62a431e596/7ec4fea00761f7a58b44d43c4f95f618beb88f9b.patch) (2216 bytes) - added by stephsolis on 2013-04-22 02:40:45Z - Addressed review suggestions
* [5780v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7a4/7a422879511b7e2652ca8839f9042a62a431e596/19c6eed058b095bf6369b0ce3a8f50cf23bdfce6.patch) (2616 bytes) - added by stephsolis on 2013-06-14 20:47:14Z - Addressed more review suggestions.

<details><summary>Searchable metadata</summary>

```
trac-id__5780 5780
type__defect defect
reporter__j2a j2a
priority__normal normal
milestone__ 
branch__branches_unbound_nick_5780 branches/unbound-nick-5780
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1342307183000000 1342307183000000
changetime__1371796115000000 1371796115000000
version__None None
owner__tom_prince tom.prince
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Pretty sure j2a meant to do this...

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)|@tenth10th set owner to @tenth10th|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to j2a|
|-|-|

Thanks.

  1. The docstrings for `test_utf8Messages` and `test_wrongUtf8Nick` do half of a good job.  They should explain the desired behavior as well as explaining the case they are exercising, though.
  1. Please separate methods from each other by two blank lines and wrap lines at less than 80 columns.
  1. `nickname = params[0]` isn't likely to raise a `UnicodeDecodeError`, right?  So the line being added inside the try/except could probably go before the `try` statement begins.  It's nice to minimize the amount of code inside try blocks.

Thanks again!  It's really great to see you working on this area of Twisted.

**Comment by twisted-trac** [comment]
|<img alt="stephsolis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephsolis set owner to stephsolis<br>stephsolis set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="stephsolis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephsolis removed owner<br>stephsolis set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [38481]) Branching to 'unbound-nick-5780'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to stephsolis|
|-|-|

Thanks for moving this patch forward, stephsolis.  Sorry for the delay on this review; we have quite a [backlog](http://twistedmatrix.com/trac/report/15) right now.

There are now some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/unbound-nick-5780), but due to some recent maintenance, many of the builders are unavailable.

A few minor points:

  1. Please make life a bit easier for the person who eventually ports this code to py3k and [[Python3#Reviewerchecklist](https://twistedmatrix.com/trac/wiki/Python3#Reviewerchecklist) use `b''` or `u''` strings], so that it will be unambiguous whether you're attempting to represent bytes or text with your strings.
  2. Manually typing in UTF-8 escapes to byte strings is unclear.  Instead, create a text (unicode) string, then encode it to UTF-8.  Practically speaking this just means "use `\u` instead of `\x` escapes", but this means it's a lot easier for someone reading the test to tell for example, if your test code points have any surrogate pairs in them.
  3. In point 1 of his previous review, when he said that the docstrings do "half of a good job", exarkun was referring to our preferred convention for test docstrings. Sadly I don't think this is documented officially anywhere, so I'll re-explain it again here.  Test docstrings should make a positive statement about the attributes of the desired behavior of the code under test.  So saying that something "does not raise an exception" is not helpful; most code doesn't raise an exception and it doesn't do a whole bunch of other things too :).  Instead, docstrings should say things like "when a UTF-8 message is sent with `sendMessage`, it will be written to the transport."  Also, if you're going to talk about exceptions, you should talk about specific types; "Exception" is a bit too vague.  (It would be silly to talk about `NameError` though.)
  4. This stuff all talks about UTF-8, but actually the encoding of a particular `IRCUser` is a settable attribute, so it should talk about the encoding of the user's connection.
  5. This one's just a quibble really; it's ["invalid" UTF-8](https://en.wikipedia.org/wiki/UTF-8#Invalid_byte_sequences), not "wrong UTF-8", so name / document the test accordingly.
  6. This change, like all changes, needs a [NEWS file](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).  I leave it to your discretion whether this should be a `bugfix` or `misc` (whether it should be described in the changelog or not, respectively).

If you were a committer, I'd say all these changes were small enough that you could just commit it yourself when you're done.  Unfortunately since you're not, I have to ask you to submit another patch to finish this.  Feel free to ping me personally to remind me to review it so we can get it merged promptly when you have responded to this review.  Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="stephsolis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephsolis removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince commented|
|-|-|

(In [38808]) Apply 5780v3.patch from stephsolis.

Refs: [#5780](https://github.com/twisted/twisted/issues/5780)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set owner to @tomprince<br>@tomprince set status to `assigned`|
|-|-|

This looks good. I'll merge after the buildbot runs.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

Actually, the newsfile says that it allows `NICK` commands with non-UTF8 data. But, it doesn't *allow* them, it just properly reports an error, for them.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince set status to `closed`|
|-|-|

(In [38864]) Merge unbound-nick-5780: Fix handling of invalid UTF-8 in IRC NICK commands.

Author: j2a, stephsolis
Reviewers: exarkun, glyph, tom.prince
Fixes: [#5780](https://github.com/twisted/twisted/issues/5780)

Previously, when sent a NICK command that had invalid UTF-8,
it would fail with an UnboundLocalError. It now properly
reports an error to the user, instead.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

Sorry for not catching this earlier but the topfile is in the wrong place and this change will only be reported in the core section of the news file, but it should be in words. It's easy to make this mistake, wonder if the commit hook/tools can be smarter about this.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince set status to `reopened`|
|-|-|

(In [38871]) Revert r38864: Topfile in the wrong place.

Reopens: [#5780](https://github.com/twisted/twisted/issues/5780)

enforcenews.py should be smarter about this.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince set status to `closed`|
|-|-|

(In [38872]) Merge unbound-nick-5780: Fix handling of invalid UTF-8 in IRC NICK commands.

Author: j2a, stephsolis
Reviewers: exarkun, glyph, tom.prince
Fixes: [#5780](https://github.com/twisted/twisted/issues/5780)

Previously, when sent a NICK command that had invalid UTF-8,
it would fail with an UnboundLocalError. It now properly
reports an error to the user, instead.
